chrome/content/zotero/xpcom/attachments.js

```diff
@@ -2241,95 +2241,195 @@ Zotero.Attachments = new function(){
+			const lastFirst = ['full-reversed', 'family-given', 'last-first'];
+			const first = ['given', 'first'];
+			const last = ['family', 'last'];
+
+			if (firstLast.includes(name)) {
+				return initializeFn(creator.firstName, ['full', ...first].includes(initialize), initializewith) + namepartseparator + initializeFn(creator.lastName, ['full', ...last].includes(initialize), initializewith);
 			}
-			
-			return str.replace(re, f);
-		}
-		
-		formatString = rpl('creator');
-		formatString = rpl('year');
-		formatString = rpl('title');
-		
+			else if (lastFirst.includes(name)) {
+				return initializeFn(creator.lastName, ['full', ...last].includes(initialize), initializewith) + namepartseparator + initializeFn(creator.firstName, ['full', ...first].includes(initialize), initializewith);
+			}
+			else if (first.includes(name)) {
+				return initializeFn(creator.firstName, ['full', ...first].includes(initialize), initializewith);
+			}
+
+			return initializeFn(creator.lastName, ['full', ...last].includes(initialize), initializewith);
+		};
+
+		const commonCreators = (value, { max = Infinity, order = 'asc', name = 'family', namepartseparator = ' ', join = ', ', initialize = '', initializewith = '.' } = {}) => {
+			return getSlicedCreatorsOfType(value, order === "desc" ? -max : max)
+				.map(c => transformName(c, { name, namepartseparator, initialize, initializewith }))
+				.join(join);
+		};
+
+		const fields = Zotero.ItemFields.getAll()
+			.map(f => f.name)
+			.reduce((obj, name) => {
+				obj[name] = (args) => {
+					return common(item.getField(name, false, true), args);
+				};
+				return obj;
+			}, {});
+
+		const year = (args) => {
+			let value = item.getField('date', true, true);
+			if (value) {
+				value = Zotero.Date.multipartToSQL(value).substr(0, 4);
+				if (value == '0000') {
+					value = '';
+				}
+			}
+			return common(value, args);
+		};
+
+		const itemType = args => common(Zotero.ItemTypes.getLocalizedString(item.itemType), args);
+		const itemTypeRaw = item.itemType;
+
+		const creatorFields = ['authors', 'editors', 'creators'].reduce((obj, name) => {
+			obj[name] = (args) => {
+				return common(commonCreators(name, args), args);
+			};
+			return obj;
+		}, {});
+
+		const firstCreator = args => common(
+			// 74492e40 adds \u2068 and \u2069 around names in the `firstCreator` field, which we don't want in the filename
+			// We might actually want to move this replacement to getValidFileName
+			item.getField('firstCreator', false, true).replaceAll('\u2068', '').replaceAll('\u2069', ''), args
+		);
+
+		const vars = { ...fields, ...creatorFields, firstCreator, itemType, itemTypeRaw, year };
+
+		formatString = Zotero.Utilities.Internal.generateHTMLFromTemplate(formatString, vars);
 		formatString = Zotero.Utilities.cleanTags(formatString);
 		formatString = Zotero.File.getValidFileName(formatString);
 		return formatString;
-	}
+	};
+
+	this.convertLegacyFormatString = function (formatString) {
```

Comment:
This is 100% overkill, given the extremely limited options we had, but very nice! There's a pref-migration mechanism in prefs.js.

Reply:
I've moved to `prefs.js` and added a migration step 7 -> 8